### PR TITLE
[WIP] fix(rust): fix credentials build failure

### LINKED ISSUE #2209: Fix Ockam credential feature
Ockam credential feature seem to be broken. Running 
```bash
cd implementations/rust/ockam/ockam

cargo check --no-default-features --features "credentials std software_vault"
```

Fails with
```bash
cargo check --no-default-features --features "credentials std software_vault"
    Checking ockam_entity v0.28.0 (ockam/implementations/rust/ockam/ockam_entity)
error[E0658]: custom discriminant values are not allowed in enums with tuple or struct variants
  --> implementations/rust/ockam/ockam_entity/src/credential/error.rs:8:37
   |
8  |     MismatchedAttributesAndClaims = 1,
   |                                     ^ disallowed custom discriminant
...
22 |     InvalidCredentialPresentation(u32),
   |     ---------------------------------- tuple variant defined here
   |
   = note: see issue #60553 <https://github.com/rust-lang/rust/issues/60553> for more information

error[E0277]: the trait bound `credential::messages::CredentialProtocolMessage: ockam_core::Message` is not satisfied
   --> implementations/rust/ockam/ockam_entity/src/credential/workers/verifier.rs:56:14
    |
56  |         msg: Routed<Self::Message>,
    |              ^^^^^^^^^^^^^^^^^^^^^ the trait `ockam_core::Message` is not implemented for `credential::messages::CredentialProtocolMessage`
    |
note: required by a bound in `Routed`
   --> /ockam/implementations/rust/ockam/ockam_core/src/message.rs:109:22
    |
109 | pub struct Routed<M: Message> {
    |                      ^^^^^^^ required by this bound in `Routed`

error[E0277]: the trait bound `credential::messages::CredentialProtocolMessage: ockam_core::Message` is not satisfied
   --> implementations/rust/ockam/ockam_entity/src/credential/workers/presenter.rs:61:14
    |
61  |         msg: Routed<Self::Message>,
    |              ^^^^^^^^^^^^^^^^^^^^^ the trait `ockam_core::Message` is not implemented for `credential::messages::CredentialProtocolMessage`
    |
note: required by a bound in `Routed`
   --> ockam/implementations/rust/ockam/ockam_core/src/message.rs:109:22
    |
109 | pub struct Routed<M: Message> {
    |                      ^^^^^^^ required by this bound in `Routed`

error[E0277]: the trait bound `credential::messages::CredentialProtocolMessage: ockam_core::Message` is not satisfied
   --> implementations/rust/ockam/ockam_entity/src/credential/workers/listener.rs:34:14
    |
34  |         msg: Routed<Self::Message>,
    |              ^^^^^^^^^^^^^^^^^^^^^ the trait `ockam_core::Message` is not implemented for `credential::messages::CredentialProtocolMessage`
    |
note: required by a bound in `Routed`
   --> ockam/implementations/rust/ockam/ockam_core/src/message.rs:109:22
    |
109 | pub struct Routed<M: Message> {
    |                      ^^^^^^^ required by this bound in `Routed`

error[E0277]: the trait bound `credential::messages::CredentialProtocolMessage: ockam_core::Message` is not satisfied
   --> implementations/rust/ockam/ockam_entity/src/credential/workers/issuer.rs:65:14
    |
65  |         msg: Routed<Self::Message>,
    |              ^^^^^^^^^^^^^^^^^^^^^ the trait `ockam_core::Message` is not implemented for `credential::messages::CredentialProtocolMessage`
    |
note: required by a bound in `Routed`
   --> ockam/implementations/rust/ockam/ockam_core/src/message.rs:109:22
    |
109 | pub struct Routed<M: Message> {
    |                      ^^^^^^^ required by this bound in `Routed`

error[E0277]: the trait bound `credential::messages::CredentialProtocolMessage: ockam_core::Message` is not satisfied
   --> implementations/rust/ockam/ockam_entity/src/credential/workers/holder.rs:68:14
    |
68  |         msg: Routed<Self::Message>,
    |              ^^^^^^^^^^^^^^^^^^^^^ the trait `ockam_core::Message` is not implemented for `credential::messages::CredentialProtocolMessage`
    |
note: required by a bound in `Routed`
   --> ockam/implementations/rust/ockam/ockam_core/src/message.rs:109:22
    |
109 | pub struct Routed<M: Message> {
    |                      ^^^^^^^ required by this bound in `Routed`

Some errors have detailed explanations: E0277, E0658.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `ockam_entity` due to 6 previous errors
```

We should fix and also add CI checks

### PR DESCRIPTION
**Pull request by wkk**
Signed-off-by: Bin Wang <wkk@users.noreply.github.com>

## Current Behaviour

As mentioned in #2209. currently building ockam with credentials features will fail. I also noticed that `ockam_examples/example_projects/credentials/` is not building properly.

## Proposed Changes

I believe this PR fixes both the credentials build (closes #2209 ) and the example program.

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://www.ockam.io/learn/how-to-guides/contributing/CONTRIBUTING#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://www.ockam.io/learn/how-to-guides/high-performance-team/conduct/).
- [x] I have accepted the Ockam [Contributor Licence Agreement](https://www.ockam.io/learn/how-to-guides/contributing/cla/) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/ockam-network/contributors/blob/master/CONTRIBUTORS.csv) file in a separate pull request to the [ockam-network/contributors](https://github.com/ockam-network/contributors) repository.

### CODE REVIEW

**Comment by wkk** [comment]
However, although the example program builds successfully, it will fail during runtime. I have traced the reason:
1. The `office` program (which acts as the Issuer) creates an issuance key

https://github.com/wkk/ockam/blob/d0132d23774f4313d7b4064b39d408b3de656fa7/implementations/rust/ockam/ockam_examples/example_projects/credentials/src/bin/office.rs#L19

2. This key has a default implementation of ed25519

https://github.com/wkk/ockam/blob/d0132d23774f4313d7b4064b39d408b3de656fa7/implementations/rust/ockam/ockam_entity/src/key_attributes.rs#L30-L39

3. The `bob` and `door` programs will request the signing public key from `office` (the Issuer). However, in the `ockam_credential` crate this key is assume to be a BBS+ public key. As the key length doesn't match the conversion will fail.

https://github.com/wkk/ockam/blob/d0132d23774f4313d7b4064b39d408b3de656fa7/implementations/rust/ockam/ockam_entity/src/credential/entity_credentials.rs#L526-L531

I'm not sure how to fix this. There seem to be some options, for example I can create another method like `create_credential_issuance_key` which generates a BBS+ key, or I can modify the create_key method to take a parameter of what type of key to create. I'm not sure which one is the best. I've looked in the proposals repo and there is no specifications about the credential protocol. Please help. Any input is welcomed

**Comment by mrinalwadhwa** [comment]
@wkk Thank you for looking into this and attempting to fix the issue. The best person to advise in how to proceed is @SanjoDeundiak but we're all taking a year end vacation so we may not hear from him until next week.

File: implementations/rust/ockam/ockam_entity/src/credential/error.rs [hunk 1]
```diff
@@ -5,7 +5,7 @@ use ockam_core::Error;
 #[derive(Clone, Copy, Debug)]
 pub enum CredentialError {
     /// Mismatched number of attributes in schema and provided claims to be signed
-    MismatchedAttributesAndClaims = 1,
+    MismatchedAttributesAndClaims,
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_entity/src/credential/error.rs:8]: @wkk could you please elaborate why do we need to change this here?

**Review by SanjoDeundiak — CHANGES_REQUESTED** [review]
Tnx for the contribution! One small question and we should be good to merge

File: implementations/rust/ockam/ockam_entity/src/credential/error.rs (hunk 1 above)
Comment (wkk) [reply implementations/rust/ockam/ockam_entity/src/credential/error.rs:8]: To use custom discriminant values with an enumeration, it's required that none of the variants of the enum has associated data/fields. `CredentialError::InvalidCredentialPresentation(u32)` has a `u32` field so if we try to compile with the custom discriminant value there will be the following error:

```
error[E0658]: custom discriminant values are not allowed in enums with tuple or struct variants
  --> implementations/rust/ockam/ockam_entity/src/credential/error.rs:8:37
   |
8  |     MismatchedAttributesAndClaims = 1,
   |                                     ^ disallowed custom discriminant
...
22 |     InvalidCredentialPresentation(u32),
   |     ---------------------------------- tuple variant defined here
   |
   = note: see issue #60553 <https://github.com/rust-lang/rust/issues/60553> for more information
```

Reference: https://doc.rust-lang.org/reference/items/enumerations.html#custom-discriminant-values-for-field-less-enumerations

Comment (wkk) [reply implementations/rust/ockam/ockam_entity/src/credential/error.rs:8]: Also I don't think the custom discriminant value is used anywhere. There is already an `as_u32` function defined for the `CredentialError` struct. So I think it should be safe to remove this.

Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_entity/src/credential/error.rs:8]: @wkk sure, you can delete it

**Comment by SanjoDeundiak** [comment]
> However, although the example program builds successfully, it will fail during runtime. I have traced the reason:
> 
> 1. The `office` program (which acts as the Issuer) creates an issuance key
> 
> https://github.com/wkk/ockam/blob/d0132d23774f4313d7b4064b39d408b3de656fa7/implementations/rust/ockam/ockam_examples/example_projects/credentials/src/bin/office.rs#L19
> 
> 2. This key has a default implementation of ed25519
> 
> https://github.com/wkk/ockam/blob/d0132d23774f4313d7b4064b39d408b3de656fa7/implementations/rust/ockam/ockam_entity/src/key_attributes.rs#L30-L39
> 
> 3. The `bob` and `door` programs will request the signing public key from `office` (the Issuer). However, in the `ockam_credential` crate this key is assume to be a BBS+ public key. As the key length doesn't match the conversion will fail.
> 
> https://github.com/wkk/ockam/blob/d0132d23774f4313d7b4064b39d408b3de656fa7/implementations/rust/ockam/ockam_entity/src/credential/entity_credentials.rs#L526-L531
> 
> I'm not sure how to fix this. There seem to be some options, for example I can create another method like `create_credential_issuance_key` which generates a BBS+ key, or I can modify the create_key method to take a parameter of what type of key to create. I'm not sure which one is the best. I've looked in the proposals repo and there is no specifications about the credential protocol. Please help. Any input is welcomed

Nice catch, I think it's fine if it doesn't work. It's in very early development state

**Comment by wkk** [comment]
Thanks. I guess this can be merged as it is? I can open a new issue tracking the status of the credentials example once this is merged.

**Review by SanjoDeundiak — APPROVED** [review]

**Comment by jared-s** [comment]
Thanks! Rebased and merged as https://github.com/ockam-network/ockam/pull/2425

**Pull request closed without merging**